Resources can overflow

If you leave the game running for a few minutes, it's quite likely that the resources will overflow their max values, since they're `int`s. This shouldn't be that hard to fix (probably just as simple as changing them to `long`s)

I'm having troubles compiling this. Does the current build work fine for you?

Also, might be a good idea to mention the version of SFML you are using. I was scratching my head for a while wondering why I was getting some of the errors I was gettting. Turns out, just like CMake isn't the latest on Ubuntu 16, SFML isn't either, it's at 2.3.2 on Ubuntu 16.

This is the current error ( among others)
solar-system/game.cpp:221:83: error: converting to ‘std::vector<std::tuple<Planet*, Planet*, float> >::value_type {aka std::tuple<Planet*, Planet*, float>}’ from initializer list would use explicit constructor ‘constexpr std::tuple< <template-parameter-1-1> >::tuple(_UElements&& ...) [with _UElements = {Planet* const&, Planet* const&, float&}; <template-parameter-2-2> = void; _Elements = {Planet*, Planet*, float}]’
            edges.push_back({edge.first.first, edge.first.second, edge.second});
                      
Thanks! :)

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -15,7 +15,7 @@ set(SOURCE_FILES
 add_executable(solar_system ${SOURCE_FILES})
 
 set(CMAKE_MODULE_PATH "${CMAKE_CURRENT_LIST_DIR}/cmake_modules")
-find_package(SFML REQUIRED system window graphics)
+find_package(SFML 2.4 REQUIRED system window graphics)
 
 if (SFML_FOUND)
     include_directories(${SFML_INCLUDE_DIR})

--- a/config.h
+++ b/config.h
@@ -131,7 +131,7 @@
 #define SHIP_DRAG N(0.25)
 
 // Ships can only carry this many of each resource
-#define MAX_CARGO_PER_RESOURCE 5000
+#define MAX_CARGO_PER_RESOURCE 5000L
 
 // Each ship consumes this much metal from the sender
 #define SHIP_METAL_USAGE 500

--- a/planet.cpp
+++ b/planet.cpp
@@ -128,13 +128,18 @@ void Planet::update(Game* game, float dt) {
 
 	// TODO have people die off if there's not enough food
     // Update Population using logistic model
-    int initialPop = resources.store[Population]; // Initial population of planet
+    long initialPop = resources.store[Population]; // Initial population of planet
     double k = 0.1;                               // Relative growth rate coefficient
     double A = (capacity - initialPop) / initialPop;
     double e = std::exp(1.0);
 
+    // Start reducing population due to shortage of food
+    if(resources.store[Food]==0){
+        k = -0.1;
+    }
+
     if (initialPop != 0) {
-        resources.store[Population] = int(capacity / (1 + A * pow(e, (-k * dt))));
+        resources.store[Population] = long(capacity / (1 + A * pow(e, (-k * dt))));
     }
 
 	// TODO: would like to implement people management ex. # of framers, scientists, engineers, 
@@ -145,27 +150,35 @@ void Planet::update(Game* game, float dt) {
     }
 
     // Update others resources stats
-    resources.store[Species] += randRange(0, 10000);
-    resources.store[Metal] += randRange(0, 10000);
-    resources.store[Wood] += randRange(0, 10000);
-    resources.store[Water] += randRange(0, 10000);
+    resources.store[Species] += randRange(-500, 10000);
+    resources.store[Metal] += randRange(-500, 10000);
+    resources.store[Wood] += randRange(-500, 10000);
+    resources.store[Water] += randRange(-500, 10000);
 
-	// If there's no people their shouldn't be any weapons
-	if (resources.store[Population] > 0) {
-		resources.store[Weaponary] = randRange(0, 10000);
-	}
+  	// If there's no people their shouldn't be any weapons
+  	if (resources.store[Population] > 0) {
+  		resources.store[Weaponary] = randRange(0, 10000);
+  	}
 
-	// Each farmer on the planet can grow food equal to the bio diversity of the planet, however each person must eats one
-  if(!resources.store[Population]) { // No farmers, no food
-      resources.store[Food] = 0;
-  } else {
-      resources.store[Food] = int((number_of_farmers*biodiversity)/std::sqrt(resources.store[Population]));
-  }
-
-    // Food cap
-    if(resources.store[Food] > foodCap) {
-        resources.store[Food] = foodCap;
+	  // Each farmer on the planet can grow food equal to the bio diversity of the planet, however each person must eats one
+    if(!resources.store[Population]) { // No farmers, no food
+        resources.store[Food] = 0;
+    } else {
+        resources.store[Food] = int((number_of_farmers*biodiversity)/std::sqrt(resources.store[Population]));
     }
+
+    // Cap on all resources
+    resources.store[Species] = max(0L, resources.store[Species]);
+    resources.store[Metal] = max(0L, resources.store[Metal]);
+    resources.store[Wood] = max(0L, resources.store[Wood]);
+    resources.store[Water] = max(0L, resources.store[Water]);
+    resources.store[Food] = max(0L, resources.store[Food]);
+
+    resources.store[Species] = min(speciesCap, resources.store[Species]);
+    resources.store[Metal] = min(metalCap, resources.store[Metal]);
+    resources.store[Wood] = min(woodCap, resources.store[Wood]);
+    resources.store[Water] = min(waterCap, resources.store[Water]);
+    resources.store[Food] = min(foodCap, resources.store[Food]);
 }
 
 sf::Vector2f Planet::getPosition(sf::Vector2f center) {

--- a/planet.h
+++ b/planet.h
@@ -20,9 +20,13 @@
 class Game;
 
 class Planet {
-    const int foodCap = 50000;
-    const int farmersCap = 5000000;
-
+    const long foodCap = 50000;
+    const long farmersCap = 5000000;
+    const long metalCap = 50000000;
+    const long woodCap = 50000000;
+    const long waterCap = 50000000;
+    const long speciesCap = 500000;
+    
     sf::CircleShape shape;
     sf::CircleShape orbit;
     sf::CircleShape border;

--- a/resource.h
+++ b/resource.h
@@ -21,7 +21,7 @@ enum ResourceType {
     Weaponary
 };
 
-typedef std::map<ResourceType, int> Store;
+typedef std::map<ResourceType, long> Store;
 
 /** Resources
  *

--- a/sidebar.cpp
+++ b/sidebar.cpp
@@ -4,7 +4,7 @@
 
 #include "sidebar.h"
 
-std::string formatNumber(int n) {
+std::string formatNumber(long long n) {
     return std::to_string(n);
 }
 
@@ -61,7 +61,7 @@ PlanetInspector::PlanetInspector(Planet *planet)
         metal = TextWidget(&body, "Metal: " + formatNumber(planet->resources.store[Metal]) + " tons", false);
         wood = TextWidget(&body, "Wood: " + formatNumber(planet->resources.store[Wood]) + " tons", false);
         water = TextWidget(&body, "Water: " + formatNumber(planet->resources.store[Water]) + " cubic feet", false);
-        food = TextWidget(&body, "Food: " + formatNumber(planet->resources.store[Food]) + " units", false);
+        food = TextWidget(&body, "Food: " + formatNumber(planet->resources.store[Food]) + " tons", false);
         weaponary = TextWidget(&body, "Weaponry: " + formatNumber(planet->resources.store[Weaponary]) + " units", false);
 
         widgets.addWidget(&resources);
@@ -130,7 +130,7 @@ void PlanetInspector::updateWidgets() {
     metal.setString("Metal: " + formatNumber(planet->resources.store[Metal]) + " tons");
     wood.setString("Wood: " + formatNumber(planet->resources.store[Wood]) + " tons");
     water.setString("Water: " + formatNumber(planet->resources.store[Water]) + " cubic feet");
-    populationResource.setString("Population: " + formatNumber(planet->resources.store[Population]) + " inhabitants");
+    //populationResource.setString("Population: " + formatNumber(planet->resources.store[Population]) + " inhabitants");
     food.setString("Food: " + formatNumber(planet->resources.store[Food]) + " tons");
     weaponary.setString("Weaponry: " + formatNumber(planet->resources.store[Weaponary]) + " units");
 }

--- a/sidebar.cpp
+++ b/sidebar.cpp
@@ -62,7 +62,7 @@ PlanetInspector::PlanetInspector(Planet *planet)
         wood = TextWidget(&body, "Wood: " + formatNumber(planet->resources.store[Wood]) + " tons", false);
         water = TextWidget(&body, "Water: " + formatNumber(planet->resources.store[Water]) + " cubic feet", false);
         food = TextWidget(&body, "Food: " + formatNumber(planet->resources.store[Food]) + " units", false);
-        weaponary = TextWidget(&body, "Weaponary: " + formatNumber(planet->resources.store[Weaponary]) + " units", false);
+        weaponary = TextWidget(&body, "Weaponry: " + formatNumber(planet->resources.store[Weaponary]) + " units", false);
 
         widgets.addWidget(&resources);
         widgets.addWidget(&metal);
@@ -132,7 +132,7 @@ void PlanetInspector::updateWidgets() {
     water.setString("Water: " + formatNumber(planet->resources.store[Water]) + " cubic feet");
     populationResource.setString("Population: " + formatNumber(planet->resources.store[Population]) + " inhabitants");
     food.setString("Food: " + formatNumber(planet->resources.store[Food]) + " tons");
-    weaponary.setString("Weaponary: " + formatNumber(planet->resources.store[Weaponary]) + " units");
+    weaponary.setString("Weaponry: " + formatNumber(planet->resources.store[Weaponary]) + " units");
 }
 
 

